---
title: "wgs_base_editing_on_off_target_analysis"
author: "Anlu Chen"
date: "Nov 9, 2020"
output: html_document
---

```{r setup}
wd<-getwd()
work_dir<-paste0(wd,"/wgs")
library(pheatmap)
library(ggplot2)
library(dplyr)
library(tibble)
library(tidyr)
library(readr)
require(purrr)
library(data.table)
library(RColorBrewer)
library(viridis)

violin_scale <- "width"
color_begin <- 0
color_end <- 0.9
sampleorder=paste0("colony_",c(4:10))
sampleorderl=paste0("colony_",c(1,4:10))
chrOrder <-c((1:22),"X","Y")
chrOrder2 <-paste0("chr",chrOrder)
```

```{r export raw reads mapped on the 41bp region of WGS samples in one file}
#######################Load reads from RNAseq############################################
# path to the data
# this file is the same as in baseG
wgspath <- Sys.glob(file.path(work_dir,"CRISPRessoWGS_on_colony_*.bam_len41_hg38_baseC","REPORT_READS_ALIGNED_TO_SELECTED_REGIONS_WGS.txt"))
# Read data, only the "n_reads" column
wgsreads <- wgspath %>%
  map(read_tsv,col_types = "________d__", col_names = TRUE)  %>%
  purrr::reduce(cbind)

##Add sample name and reorder samples
filename<-as.data.frame(dirname(wgspath))
colnames(filename)<-"path"
spname<-filename%>%
  separate(path,c(NA,NA,NA,NA,NA,"spnamel"),sep = "/", extra = "drop")%>%
  separate(spnamel,c(NA,NA,"spnames","index"),sep = "_", extra = "drop") %>%
  mutate(spnamess=paste0(spnames,"_",index))%>%
  separate(spnamess,c("spname",NA),sep = "[.]", extra = "drop") %>% #sep = "." doesn't work 
  pull(spname)
colnames(wgsreads)<-spname
wgsreads<-wgsreads[,sampleorderl]


#Rename sample name
colnames(wgsreads)<-c("NC",paste0("CLONE ",c("19","21","19-1","19-16","19-21")))

# Add genelist to the data
aa<-wgspath[1] %>%
  # read in all the files, appending the path before the filename
  map(read_tsv,col_types = "c__________________", col_names = TRUE)
row.names(wgsreads) <-unlist(aa) 
wgsreads<- wgsreads[order(row.names(wgsreads)), ]

write.csv(wgsreads,file.path(work_dir,"wgs_raw_reads_mapped_on_region_wgs.csv"),row.names = T)
```



```{r modified read%}
# path to the data
data_path <- Sys.glob(file.path(work_dir,"CRISPRessoWGS_on_colony_*.bam_len41_hg38","SAMPLES_QUANTIFICATION_SUMMARY.txt"))   
# Read data, only the "Modified%" column
hpr <- data_path %>%
  # read in all the files, appending the path before the filename
  map(read_tsv,col_types = "__d________________", col_names = TRUE)%>%
  # lapply("[", "Modified%") %>%
  purrr::reduce(cbind)

##Add sample name and reorder samples
filename<-as.data.frame(dirname(data_path))
colnames(filename)<-"path"
spname<-filename%>%
  separate(path,c(NA,NA,NA,NA,NA,"spnamel"),sep = "/", extra = "drop")%>%
  separate(spnamel,c(NA,NA,"spnames","index"),sep = "_", extra = "drop") %>%
  mutate(spnamess=paste0(spnames,"_",index))%>%
  separate(spnamess,c("spname",NA),sep = "[.]", extra = "drop") %>% #sep = "." doesn't work 
  pull(spname)
colnames(hpr)<-spname
hpr<-hpr[,sampleorderl]


#Rename sample name
colnames(hpr)<-c("NC",paste0("CLONE ",c("19","21","19-1","19-16","19-21")))

#Add gene name
aa<-data_path[1] %>%
  # read in all the files, appending the path before the filename
  map(read_tsv,col_types = "c__________________", col_names = TRUE)
row.names(hpr) <-unlist(aa) 
hpr<-hpr[order(row.names(hpr)), ]

#remove NA rows
hpr<-hpr[complete.cases(hpr), ]

#Make a matrix as input
hpr<-as.matrix.data.frame(hpr)

##Plot
p2<-pheatmap(hpr,
         col= colorRampPalette(brewer.pal(10, "Reds"))(100),
         border="white", 
         main="Heatmap on total edits% based on WGS result",
         cluster_cols = F, 
         cluster_rows = F,
         show_rownames = T,
         show_colnames = T,
         legend = T, 
         fontsize_row = 10, 
         fontsize_col = 10,
         angle_col = 45)

print(paste0("In total, ",nrow(hpr)," targets in the analysis"))  

#Save the heatmap
ggsave("wgs_total_edits_plot_red.png",p2,"png",file.path(work_dir),height = 10, width = 7, units = "in")
ggsave("wgs_total_edits_plot_red.pdf",p2,"pdf",file.path(work_dir),height = 10, width = 7, units = "in")

```


```{r percentage of edited on-targets}
basecpath <- Sys.glob(file.path(work_dir,"*_baseC","*","Selected_nucleotide_percentage_table_around_sgRNA_*.txt"))

# Read data, only the "T" column to calculate C to T mutation%
cpcs<- basecpath %>%
  map(read_tsv,col_names = T)  %>%
  bind_rows() %>%
  filter(X1=="T") 
cpcs<-cpcs[,-1] 
col_order <- paste0("C",1:30)
#order column C1 to C30
cpcs <- cpcs[, col_order]
# Calculate the presentage of successful C to T edits in %
cpcs<-round(cpcs*100,2)


filename<-as.data.frame(dirname(basecpath))
colnames(filename)<-"path"
cpcsf<-filename%>%
  separate(path,c(NA,NA,NA,NA,NA,"spnamel","regnamel"),sep = "/", extra = "drop")%>%
  separate(spnamel,c(NA,NA,"spnames","index"),sep = "_", extra = "drop") %>%
  mutate(spnamess=paste0(spnames,"_",index))%>%
  separate(spnamess,c("spname",NA),sep = "[.]", extra = "drop") %>% #sep = "." doesn't work 
  separate(regnamel,c(NA,NA,"regname"),sep = "_", extra = "merge") %>% ##extra=merge to keep multiple targets on the same gene
  select(regname,spname)%>%
  cbind(cpcs)

rm(filepath,col_order)


basegpath <- Sys.glob(file.path(work_dir,"*_baseG","*","Selected_nucleotide_percentage_table_around_sgRNA_*.txt"))

# Read data, only the "A" column to calculate G to A mutation%
cpgs<- basegpath %>%
  map(read_tsv,col_names = T)  %>%
  bind_rows() %>%
  filter(X1=="A") 
cpgs<-cpgs[,-1] 
col_order <- paste0("G",1:30)
cpgs <- cpgs[, col_order]
# Calculate the presentage of successful C to T edits in %
cpgs<-round(cpgs*100,2)

filename<-as.data.frame(dirname(basegpath))
colnames(filename)<-"path"
cpgsf<-filename%>%
  separate(path,c(NA,NA,NA,NA,NA,"spnamel","regnamel"),sep = "/", extra = "drop")%>%
  separate(spnamel,c(NA,NA,"spnames","index"),sep = "_", extra = "drop") %>%
  mutate(spnamess=paste0(spnames,"_",index))%>%
  separate(spnamess,c("spname",NA),sep = "[.]", extra = "drop") %>% #sep = "." doesn't work 
  separate(regnamel,c(NA,NA,"regname"),sep = "_", extra = "merge") %>% ##extra=merge to keep multiple targets on the same gene
  select(regname,spname)%>%
  cbind(cpgs)

cpcgsf<-cbind(cpcsf,cpgsf)

#Remove the duplicate egname,spname columns
cpcgsf<-cpcgsf[,-c(33:34)]
cpcgsf<-cpcgsf%>%
    arrange(regname,factor(spname,levels=sampleorderl))  #reorder based on regname and DEFNED sample order

write.csv(cpcgsf,file.path(work_dir,"wgs_Base_edit_all_CtoT_GtoA_precentage_in_region.csv") ,row.names = F)  
saveRDS(cpcgsf,file.path(work_dir,"wgs_Base_edit_all_CtoT_GtoA_precentage_in_region.rds"))


###########################################################################################
##read file with C or G location for each target
tgtc<-read.csv(file.path(work_dir,"Target_tag_centered_minus15.csv"))


# cpcgsf<-readRDS(file.path(work_dir,"wgs_Base_edit_all_CtoT_GtoA_precentage_in_region.rds"))  
 

newtgts<-newtgt%>%
  filter(tgtname %in% cpcgsf$regname)

cpcgsfinal<-NULL
for (i in c(1:nrow(newtgts))) {
   tmp<-cpcgsf%>%
     filter(regname==newtgts$tgtname[i])%>%
       select(regname,spname,newtgts$newtag[i])
       names(tmp)[3] <- "on_target_editting"
   cpcgsfinal<-rbind(cpcgsfinal,tmp)
}

#melt down tableframe
cpcgsop<-reshape2::dcast(cpcgsfinal, regname~spname)
#round Modified% to 2 decimal places
cpcgsop<-cpcgsop%>%
  mutate_if(is.numeric, round, digits=2)  
#Rename sample name
rownames(cpcgsop)<-cpcgsop$regname
cpcgsop<-cpcgsop[,-1]
cpcgsop<-cpcgsop[,c(paste0("colony_",c(1,5,7:10)))]
colnames(cpcgsop)<-c("NC",paste0("CLONE ",c("19","21","19-1","19-16","19-21")))

write.csv(cpcgsop,file.path(work_dir,"wgs_Base_edit_on_target_CtoT_GtoA_precentage.csv"), row.names = T)  
saveRDS(cpcgsop,file.path(work_dir,"wgs_Base_edit_on_target_CtoT_GtoA_precentage.rds"))
```


```{r wgs plot on-target base change}
ontgt <- readRDS(file.path(work_dir,"wgs_Base_edit_on_target_CtoT_GtoA_precentage.rds"))

#remove NA rows
ontgt<-ontgt[complete.cases(ontgt), ]

#Make a matrix as input
ontgt<-as.matrix.data.frame(ontgt)

p3<-pheatmap(ontgt,
         col= colorRampPalette(brewer.pal(10, "Reds"))(100),
         border="white", 
         main="Heatmap on on-target base edits based on WGS result",
         cluster_cols = F, 
         cluster_rows = F,
         show_rownames = T,
         show_colnames = T,
         legend = T, 
         fontsize_row = 8, 
         fontsize_col = 8,
         angle_col = 45)

#Save the heatmap
ggsave("wgs_on_target_percentage.png",p3,"png",file.path(work_dir),height = 10, width = 7, units = "in")
```



```{r off target snvs between mutect2 and strelka2}
wd<-getwd()
work_dir<-paste0(wd,"/wgs/shared")
offtgtpath <- Sys.glob(file.path(work_dir,"colony_*_shared_snvs_final.txt"))


finalsnv<-NULL

for (i in 1: length(offtgtpath)) {
data<-read.table(offtgtpath[i],header = T)

data<-data%>%
  mutate(spnamel=basename(offtgtpath[i]))%>%
  separate(spnamel,c("spnames","index"),sep = "_", extra = "drop",remove=T) %>%
  mutate(spname=paste0(spnames,"_",index))
data$spnames<-NULL
data$index<-NULL


data$strelka2_af<-NULL
strelka2_af_tmp<-NULL
  for (j in 1: nrow(data)) {
  alt_cnts_tmp <- data[j,colnames(data) %in% data$alt[j]] #change name if also include mutect2 snps
  ref_cnts_tmp <- data[j,colnames(data) %in% data$ref[j]]
  strelka2_af_tmp[j] <- round(alt_cnts_tmp/(alt_cnts_tmp + ref_cnts_tmp),3)
  }
  data$strelka2_af<-strelka2_af_tmp
  
  
  
  finalsnv<-rbind(finalsnv,data)
  rm(data,strelka2_af_tmp, alt_cnts_tmp,ref_cnts_tmp)
}
write.csv(finalsnv,file.path(work_dir,"wgs_somatic_total_snvs_mutect2_strelka2.csv"),row.names = F)
saveRDS(finalsnv,file.path(work_dir,"wgs_somatic_total_snvs_mutect2_strelka2.rds"))



```
```{r generate input file for annovar}
finalsnv<-readRDS(file.path(work_dir, "wgs_somatic_total_snvs_mutect2_strelka2.rds"))
finalsnva<-finalsnv%>%
  mutate(end=pos)%>%
  select(chr,pos,end,everything())
write.table(finalsnva,file.path(work_dir,"wgs_somatic_total_snvs_mutect2_strelka2.avinput"),row.names = F,col.names = F, quote = F, sep = "\t")


finalindel<-readRDS(file.path(work_dir, "wgs_somatic_total_indels_mutect2_strelka2.rds"))
finalindela<-finalindel%>%
  mutate(new_start=pos+1)%>%
  mutate(new_end=ifelse(nchar(as.character(alt))>nchar(as.character(ref)),pos+1,pos+nchar(as.character(ref))-1))%>%
  select(chr,new_start,new_end,ref,alt,everything())
write.table(finalindela,file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2.avinput"),row.names = F,col.names = F, quote = F, sep = "\t")




```

```{r add annovar annoation to the file}
snvanno<-read.csv(file.path(work_dir, "wgs_somatic_total_snvs_annovar.hg38_multianno.csv"),header=T)
finalsnv<-readRDS(file.path(work_dir, "wgs_somatic_total_snvs_mutect2_strelka2.rds"))
finalsnvanno<-cbind(finalsnv,snvanno)
write.csv(finalsnvanno,file.path(work_dir,"wgs_somatic_total_snvs_mutect2_strelka2_annovar.csv"),row.names = F)
saveRDS(finalsnvanno,file.path(work_dir,"wgs_somatic_total_snvs_mutect2_strelka2_annovar.rds"))

indelanno<-read.csv(file.path(work_dir, "wgs_somatic_total_indels_annovar.hg38_multianno.csv"),header=T)
finalindel<-readRDS(file.path(work_dir, "wgs_somatic_total_indels_mutect2_strelka2.rds"))
finalindelanno<-cbind(finalindel,indelanno)
write.csv(finalindelanno,file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2_annovar.csv"),row.names = F)
saveRDS(finalindelanno,file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2_annovar.rds"))

```


```{r total SNPs ploting}
wd<-getwd()
work_dir<-paste0(wd,"/wgs/shared")
finalsnvanno<-readRDS(file.path(work_dir,"wgs_somatic_total_snvs_mutect2_strelka2_annovar.rds"))

finalsnvanno<-finalsnvanno%>%
  dplyr::arrange(factor(spname,levels=sampleorder))

test<-finalsnvanno%>%
  group_by(spname)%>%
  summarise(n())  ##max counts in colony_7 70356


#Rename sample name
finalsnvanno$spname <- recode_factor(finalsnvanno$spname, colony_5 = "CLONE19", colony_7 = "CLONE21",colony_8 = "CLONE19-1",colony_9 = "CLONE19-16",colony_10 = "CLONE19-21")
#colony_6 = "CLONE11"

##reorder chromosomes
chrOrder <-c((1:22),"X","Y")
finalsnvanno<-finalsnvanno%>%
  mutate(CHR=str_replace(chr, "chr", ""))%>%
  mutate(CHR=ordered(CHR,levels=chrOrder))

#SNVs counts across samples
p5<-finalsnvanno %>%
  ggplot((aes(x = spname)))+
  geom_bar((aes(x = spname)), color="darkblue", fill="lightblue") +
  labs(title= "all SNVs across the genome",x="samples",y="Counts") +
  ylim(0,80000) +
  geom_text(stat='count', aes(label=..count..), vjust=-1) +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_snvs_mutect2_strelka2_count.png"),p5,"png",units="in")

#alternative ratio among samples
p5.1<-
  finalsnvanno%>%
  ggplot(aes(x = spname, y =mutect2_af)) +
  geom_jitter(aes(x = spname, y = mutect2_af),color="darkblue", fill="lightblue",alpha=0.8, size=0.5) +
  facet_grid() +
  scale_y_continuous() +
  labs(title= "SNVs mutation ratio",x="samples",y="SNVs mutation ratio \nnomalized to control") +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_SNVs_mutect2_strelka2_editing_ratio_based_on_mutect2_jitter.png"),p5.1,"png",units="in")


p5.2<-
  finalsnvanno%>%
  ggplot(aes(x = spname, y =mutect2_af)) +
  geom_violin(aes(x = spname, y =mutect2_af),scale = violin_scale,position=position_dodge(0.9),color="darkblue", fill="lightblue") +
  geom_boxplot(width = 0.2)+
  stat_summary(fun = median, geom = 'point', size = 1, color = 'red') +
  facet_grid() +
  labs(title= "SNVs mutation ratio",x="samples",y="SNVs mutation ratio \nnomalized to control") +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_SNVs_mutect2_strelka2_editing_ratio_based_on_mutect2_violin.png"),p5.2,"png",units="in")


p5.3<-
  finalsnvanno%>%
  ggplot(aes(x = spname, y =strelka2_af)) +
  geom_jitter(aes(x = spname, y = strelka2_af),color="darkblue", fill="lightblue",alpha=0.8, size=0.5) +
  facet_grid() +
  scale_y_continuous() +
  labs(title= "SNVs mutation ratio",x="samples",y="SNVs mutation ratio \nnomalized to control") +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_SNVs_mutect2_strelka2_editing_ratio_based_on_strelka2_jitter.png"),p5.3,"png",units="in")


p5.4<-
  finalsnvanno%>%
  ggplot(aes(x = spname, y =strelka2_af)) +
  geom_violin(aes(x = spname, y =strelka2_af),scale = violin_scale,position=position_dodge(0.9),color="darkblue", fill="lightblue") +
  geom_boxplot(width = 0.2)+
  stat_summary(fun = median, geom = 'point', size = 1, color = 'red') +
  facet_grid() +
  labs(title= "SNVs mutation ratio",x="samples",y="SNVs mutation ratio \nnomalized to control") +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_SNVs_mutect2_strelka2_editing_ratio_based_on_strelka2_violin.png"),p5.4,"png",units="in")


#alternative ratio among samples, break down into chromosomes

p5.5<-
  finalsnvanno%>%
  ggplot(aes(x = CHR, y = mutect2_af)) +
  geom_violin(aes(x = CHR, y = mutect2_af),color="darkblue", fill="lightblue") +
  geom_boxplot(width = 0.2)+
  stat_summary(fun = median, geom = 'point', size = 1, color = 'red') +
  scale_y_continuous() +
  labs(title= "SNVs mutation ratio",x=NULL,y="SNVs mutation ratio \nnomalized to control") +
  theme_bw(base_size = 10) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5)) +
  facet_wrap(~ spname,ncol = 1)
ggsave(file.path(work_dir,"wgs_somatic_total_SNVs_mutect2_strelka2_editing_ratio_chromosome_based_on_mutect2_violin.png"),p5.5,"png",units="in")

p5.6<-
  finalsnvanno%>%
  ggplot(aes(x = CHR, y = strelka2_af)) +
  geom_violin(aes(x = CHR, y = strelka2_af),color="darkblue", fill="lightblue") +
  geom_boxplot(width = 0.2)+
  stat_summary(fun = median, geom = 'point', size = 1, color = 'red') +
  scale_y_continuous() +
  labs(title= "SNVs mutation ratio",x=NULL,y="SNVs mutation ratio \nnomalized to control") +
  theme_bw(base_size = 10) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5)) +
  facet_wrap(~ spname,ncol = 1)
ggsave(file.path(work_dir,"wgs_somatic_total_SNVs_mutect2_strelka2_editing_ratio_chromosome_based_on_strelka2_violin.png"),p5.6,"png",units="in")



p5.7<-finalsnvanno %>%
  mutate(geneannotate=ifelse((Func.refGene=="exonic" |Func.refGene=="exonic;splicing"),"exonic","others"))%>%
  ggplot((aes(x = spname,fill=geneannotate)))+
 geom_bar(aes(x = spname,fill=geneannotate),position="stack") +
  labs(title= "all SNVs across the genome",x="samples",y="Counts") +
  ylim(0,80000) +
  geom_text(aes(label=..count..), stat='count',size = 3,position = position_stack(vjust = 0.5))+
  # geom_text(stat='count', aes(label=..count..), vjust=-1, position = position_stack()) +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_SNVs_mutect2_strelka2_geneannotation_simple.png"),p5.7,"png",units="in")


p5.8<-finalsnvanno %>%
  mutate(geneannotate=ifelse((Func.refGene=="exonic" |Func.refGene=="exonic;splicing"),"exonic","others"))%>%
  filter(geneannotate=="exonic")%>%
    mutate(exonannotate=ifelse((ExonicFunc.refGene=="." |ExonicFunc.refGene=="unknown" |ExonicFunc.refGene=="synonymous SNV"),"others",as.character(ExonicFunc.refGene)))%>%
  ggplot((aes(x = spname,fill=exonannotate)))+
 geom_bar(aes(x = spname,fill=exonannotate),position="stack") +
  labs(title= "all exonic SNVs across the genome",x="samples",y="Counts") +
  # ylim(0,80000) +
  geom_text(aes(label=..count..), stat='count',size = 3,position = position_stack(vjust = 0.4))+
  # geom_text(stat='count', aes(label=..count..), vjust=-1, position = position_stack()) +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_exonic_SNVs_mutect2_strelka2_exonannotation_simple.png"),p5.8,"png",units="in")



finalsnvanno<-finalsnvanno%>%
  mutate(snptype=paste0(ref,">",alt))

pie_colors <- viridisLite::plasma(n = length(unique(finalsnvanno$snptypetmp)), begin = color_begin, end = color_end)


p5.9<-finalsnvanno%>%
    ggplot((aes(x = spname,fill=snptype)))+
 geom_bar(aes(x = spname,fill=snptype),position="fill") +
  labs(title= "all SNVs across the genome: SNV type",x="samples",y="Counts") +
  # ylim(0,80000) +
  scale_fill_manual(values = pie_colors)+
  # geom_text(aes(label=..count..), stat='count',size = 3,position = position_stack(vjust = 0.4))+
  # geom_text(stat='count', aes(label=..count..), vjust=-1, position = position_stack()) +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_SNVs_mutect2_strelka2_snvtype.png"),p5.9,"png",units="in")

library(ggpubr)
p5.10<-
  finalsnvanno%>%
  ggscatter("strelka2_af","mutect2_af", 
              add = "reg.line", conf.int = TRUE, 
              cor.coef = TRUE, cor.method = "pearson",
              xlab = "VAF for each mutaion for strelka2", ylab = "VAF for each mutaion for mutect2",title="VAF of SNVs between strelka2 and mutect2")
ggsave(file.path(work_dir,"wgs_somatic_total_SNVs_mutect2_strelka2_editing_ratio_correlation.png"),p5.10,"png",units="in")


##essential genes
esl<-read.table(file.path(wd,"essential_gene.txt"))
colnames(esl)<-"longname"
esg<-esl%>%
  separate(longname,c("chr","genesymbol","essential","tagsite"),sep=",",remove=T)%>%
  filter(essential=="True")
  pull(genesymbol)   ##1947 essential transcripts 

length(unique(esg)) #1738 unique essential genes


##On targets genes
otl<-read.table(file.path(wd,"wgs","yc293_52_wgs_input.txt"))
otg<-as.data.frame(otl[,4])
colnames(otg)<-"genenamel"
otg<-otg%>%
  separate(genenamel,"gene",sep="_",extra="drop")%>%
  pull(gene)


## List of Exonic SNVs in essential gene
ees<-finalsnvanno%>%
  separate(Gene.refGene,c("Gene.refGene.A","Gene.refGene.B","Gene.refGene.C","Gene.refGene.D","Gene.refGene.E"),sep=";",remove=T)%>%
  filter(Gene.refGene.A  %in% esg|Gene.refGene.B  %in% esg|Gene.refGene.C  %in% esg|Gene.refGene.D  %in% esg|Gene.refGene.E  %in% esg)%>%  ##1943 essential genes
  filter(Func.refGene=="exonic" |Func.refGene=="exonic;splicing")
write.csv(ees,file.path(work_dir,"wgs_Exonic_SNVs_in_essential_gene.csv"), row.names = F)

## List of Exonic SNVs in essential gene, without the ones in target gene
eess<-ees%>%
  filter(!Gene.refGene.A  %in% otg & !Gene.refGene.B  %in% otg & !Gene.refGene.C  %in% otg & !Gene.refGene.D  %in% otg & !Gene.refGene.E  %in% otg)
write.csv(ees,file.path(work_dir,"wgs_Exonic_SNVs_in_essential_gene_not_on_targeted_gene.csv"), row.names = F)
  
p5.11<-
  ees%>%
  ggplot((aes(x = spname)))+
  geom_bar((aes(x = spname)), color="darkblue", fill="lightblue") +
  labs(title= "all exonic SNVs in essential gene across the genome",x="samples",y="Counts") +
  ylim(0,100) +
  geom_text(stat='count', aes(label=..count..), vjust=-1) +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_exonic_SNVs_mutect2_strelka2_in_essential_gene_count.png"),p5.11,"png",units="in")


#exclude SNVs on target genes
p5.12<-
  eess%>%  
  ggplot((aes(x = spname)))+
  geom_bar((aes(x = spname)), color="darkblue", fill="lightblue") +
  labs(title= "all exonic SNVs in essential gene across the genome",x="samples",y="Counts") +
  ylim(0,100) +
  geom_text(stat='count', aes(label=..count..), vjust=-1) +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_exonic_SNVs_mutect2_strelka2_in_essential_but_not _targeted_gene_count.png"),p5.12,"png",units="in")


p5.13<-
  ees%>%
  ggplot((aes(x = CHR)))+
  geom_bar(color="darkblue", fill="lightblue") +
  labs(title= "all exonic SNVs in essential gene across the genome",x="samples",y="Counts") +
  # ylim(0,15) +
  geom_text(stat='count', aes(label=..count..), vjust=-1, size=2,position = position_dodge(width=0.9)) +
  theme_bw(base_size = 10) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5)) +
  facet_wrap(~ spname,ncol = 1)
ggsave(file.path(work_dir,"wgs_somatic_total_exonic_SNVs_mutect2_strelka2_in_essential_gene_count_chr_breakdown.png"),p5.13,"png",units="in")


p5.14<-
  eess%>%
  ggplot((aes(x = CHR)))+
  geom_bar(color="darkblue", fill="lightblue") +
  labs(title= "all exonic SNVs in essential gene across the genome",x="samples",y="Counts") +
  # ylim(0,15) +
  geom_text(stat='count', aes(label=..count..), vjust=-1, size=2,position = position_dodge(width=0.9)) +
  theme_bw(base_size = 10) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5)) +
  facet_wrap(~ spname,ncol = 1)
ggsave(file.path(work_dir,"wgs_somatic_total_exonic_SNVs_mutect2_strelka2_in_essential_not_targeted_gene_chr_breakdown.png"),p5.14,"png",units="in")

```




```{r off target indels between mutect2 and strelka2}
wd<-getwd()
work_dir<-paste0(wd,"/wgs/shared")
offtgtpath <- Sys.glob(file.path(work_dir,"colony_*_shared_indels_final.txt"))


finalindel<-NULL

for (i in 1: length(offtgtpath)) {
data<-read.table(offtgtpath[i],header = T)

data<-data%>%
  mutate(spnamel=basename(offtgtpath[i]))%>%
  separate(spnamel,c("spnames","index"),sep = "_", extra = "drop",remove=T) %>%
  mutate(spname=paste0(spnames,"_",index))%>%
  mutate(strelka2_af=round(alt_cnts/(alt_cnts + ref_cnts),3))
data$spnames<-NULL
data$index<-NULL

  finalindel<-rbind(finalindel,data)
  rm(data)
}

write.csv(finalindel,file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2.csv"),row.names = F,quote = F)
saveRDS(finalindel,file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2.rds"))

```


```{r indels ploting}
work_dir=getwd()
work_dir<-paste0(work_dir,"/wgs/shared")
offtgts<-readRDS(file.path(work_dir, "wgs_somatic_total_indels_mutect2_strelka2_annovar.rds"))

offtgts<-offtgts%>%
  dplyr::arrange(factor(spname,levels=sampleorder))

test<-offtgts%>%
  group_by(spname)%>%
  summarise(n()) 

#Rename sample name
offtgts$spname <- recode_factor(offtgts$spname, colony_5 = "CLONE19", colony_6 = "CLONE11",colony_7 = "CLONE21",colony_8 = "CLONE19-1",colony_9 = "CLONE19-16",colony_10 = "CLONE19-21")


#indels counts across samples
p6<-offtgts %>%
  ggplot((aes(x = spname)))+
  geom_bar((aes(x = spname)), color="darkblue", fill="lightblue") +
  labs(title= "all indels across the genome",x="samples",y="Counts") +
  # ylim(0,12000) +  #different from SNPs
  geom_text(stat='count', aes(label=..count..), vjust=-1) +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2_count.png"),p6,"png",units="in")

#alternative ratio among samples
p6.1<-
  offtgts%>%
  ggplot(aes(x = spname, y =mutect2_af)) +
  geom_jitter(aes(x = spname, y = mutect2_af),color="darkblue", fill="lightblue",alpha=0.8, size=0.5) +
  facet_grid() +
  scale_y_continuous() +
  labs(title= "indels mutation ratio",x="samples",y="indels mutation ratio \nnomalized to control") +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2_editing_ratio_based_on_mutect2_jitter.png"),p6.1,"png",units="in")


p6.2<-
  offtgts%>%
  ggplot(aes(x = spname, y =mutect2_af)) +
  geom_violin(aes(x = spname, y =mutect2_af),scale = violin_scale,position=position_dodge(0.9),color="darkblue", fill="lightblue") +
  geom_boxplot(width = 0.2)+
  stat_summary(fun = median, geom = 'point', size = 1, color = 'red') +
  facet_grid() +
  labs(title= "indels mutation ratio",x="samples",y="indels mutation ratio \nnomalized to control") +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2_editing_ratio_based_on_mutect2_violin.png"),p6.2,"png",units="in")


p6.3<-
  offtgts%>%
  ggplot(aes(x = spname, y =strelka2_af)) +
  geom_jitter(aes(x = spname, y = strelka2_af),color="darkblue", fill="lightblue",alpha=0.8, size=0.5) +
  facet_grid() +
  scale_y_continuous() +
  labs(title= "indels mutation ratio",x="samples",y="indels mutation ratio \nnomalized to control") +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2_editing_ratio_based_on_strelka2_jitter.png"),p6.3,"png",units="in")


p6.4<-
  offtgts%>%
  ggplot(aes(x = spname, y =strelka2_af)) +
  geom_violin(aes(x = spname, y =strelka2_af),scale = violin_scale,position=position_dodge(0.9),color="darkblue", fill="lightblue") +
  geom_boxplot(width = 0.2)+
  stat_summary(fun = median, geom = 'point', size = 1, color = 'red') +
  facet_grid() +
  labs(title= "indels mutation ratio",x="samples",y="indel mutation ratio \nnomalized to control") +
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2_editing_ratio_based_on_strelka2_violin.png"),p6.4,"png",units="in")


#alternative ratio among samples, break down into chromosomes
chrOrder <-c((1:22),"X","Y")
offtgts<-offtgts%>%
  mutate(CHR=str_replace(chr, "chr", ""))%>%
  mutate(CHR=ordered(CHR,levels=chrOrder))

p6.5<-
  offtgts%>%
  ggplot(aes(x = CHR, y = mutect2_af)) +
  geom_violin(aes(x = CHR, y = mutect2_af),color="darkblue", fill="lightblue") +
  geom_boxplot(width = 0.2)+
  stat_summary(fun = median, geom = 'point', size = 1, color = 'red') +
  scale_y_continuous() +
  labs(title= "indels mutation ratio",x=NULL,y="indels mutation ratio \nnomalized to control") +
  theme_bw(base_size = 10) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5)) +
  facet_wrap(~ spname,ncol = 1)
ggsave(file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2_editing_ratio_chromosome_based_on_mutect2_violin.png"),p6.5,"png",units="in")

p6.6<-
  offtgts%>%
  ggplot(aes(x = CHR, y = strelka2_af)) +
  geom_violin(aes(x = CHR, y = strelka2_af),color="darkblue", fill="lightblue") +
  geom_boxplot(width = 0.2)+
  stat_summary(fun = median, geom = 'point', size = 1, color = 'red') +
  scale_y_continuous() +
  labs(title= "indels mutation ratio",x=NULL,y="indels mutation ratio \nnomalized to control") +
  theme_bw(base_size = 10) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5)) +
  facet_wrap(~ spname,ncol = 1)
ggsave(file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2_editing_ratio_chromosome_based_on_strelka2_violin.png"),p6.6,"png",units="in")

p6.7<-
  offtgts %>%
  mutate(geneannotate=ifelse((Func.refGene=="exonic" |Func.refGene=="exonic;splicing"),"exonic","others"))%>%
  ggplot((aes(x = spname,fill=geneannotate)))+
 geom_bar(aes(x = spname,fill=geneannotate),position="stack") +
  labs(title= "all indels across the genome",x="samples",y="Counts") +
  geom_text(aes(label=..count..), stat='count',size = 3,position = position_stack(vjust = 0.5))+
  theme_bw(base_size = 16) +
  theme(panel.grid.major = element_blank(), #remove both x axis and y axis grid
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"), #bold frame color as black
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        plot.title = element_text(hjust = 0.5))
ggsave(file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2_geneannotation_simple.png"),p6.7,"png",units="in")

library(ggpubr)
p6.8<-
  offtgts %>%
  ggscatter("strelka2_af","mutect2_af", 
              add = "reg.line", conf.int = TRUE, 
              cor.coef = TRUE, cor.method = "pearson",
              xlab = "VAF for each mutaion for strelka2", ylab = "VAF for each mutaion for mutect2",title="VAF of indels between strelka2 and mutect2")
ggsave(file.path(work_dir,"wgs_somatic_total_indels_mutect2_strelka2_editing_ratio_correlation.png"),p6.8,"png",units="in")

```
